src/gluonts/zebras/_freq.py

```diff
@@ -0,0 +1,179 @@
+    "M": "M",
+    "h": "H",
+    "m": "T",
+    "s": "S",
+}
+
+_freq_pandas_to_numpy = dict(
+    {
+        pd_freq: (np_freq, 1)
+        for np_freq, pd_freq in _freq_numpy_to_pandas.items()
+    },
+    **{
+        "A": ("Y", 1),
+        "AS": ("Y", 1),
+        "YS": ("Y", 1),
+        "MS": ("M", 1),
+        "MIN": ("m", 1),
+        "Q": ("M", 3),
+        "QS": ("M", 3),
+        "B": ("D", 1),
+        "W": ("D", 1),
+    },
+)
+
+
+def _canonical_name(name: str) -> str:
+    return {"MIN": "T", "Y": "A"}.get(name, name)
+
+
+@dataclass
+class Freq:
+    """
+    A class representing frequencies, such as n-days.
+
+    Note: Use ``freq`` to construct instances of ``Freq``.
+
+    We use frequency aliases from pandas over frequency names defined by numpy.
+    For example, the name for minutely is either "min" or "T", while "m"
+    and "M" represent monthly frequencies. In contrast numpy uses "m" for
+    minutely and "M" for monthly. In addition, pandas defines some frequencies
+    which do not exist in numpy, for example quarterly frequencies.
+
+    However, internally we use ``numpy.datetime64`` objects and thus we must
+    support numpy's frequency names as well. To do this we generally use base
+    frequencies (multiple = 1), since numpy otherwise aligns timestamps for us
+    which we don't want.
+
+    Weekly frequency needs to be handled specially, since numpy counts the
+    number of weeks since Thu Jan 1 1970 and uses Thursday and aligns the
+    timestamp to Thursday. We therefore use daily frequency internally and
+    align to Monday.
+    """
+
+    name: str
+    n: int
+
+    def __post_init__(self):
+        self.name = _canonical_name(self.name)
+
+    @property
+    def __init_passed_kwargs__(self):
+        return {"name": self.name, "n": self.n}
+
+    @property
+    def np_freq(self) -> NpFreq:
+        return _freq_pandas_to_numpy[self.name]
+
+    @classmethod
+    def __get_validators__(cls):
+        # pydantic support
+        yield freq
+
+    @classmethod
+    def from_pandas(cls, freq) -> Freq:
+        if not isinstance(freq, str):
+            if hasattr(freq, "freqstr"):
+                freq = freq.freqstr
+            else:
+                raise ValueError(f"Invalid freq {freq}: {type(freq)}")
+
+        match = maybe.expect(
+            re.match(r"(?P<n>\d+)?\s*(?P<freq>(?:\w|\-)+)", freq),
+            f"Unsupported freq format {freq}",
+        )
+        groups = match.groupdict()
+
+        name = groups["freq"].upper().split("-")[0]
+        n = maybe.map_or(groups["n"], int, 1)
+
+        return cls(name, n)
+
+    def to_pandas(self):
+        from pandas.tseries.frequencies import to_offset
+
+        return to_offset(str(self))
+
+    def shift(self, start: np.datetime64, count: int) -> np.datetime64:
+        if self.name == "B":
+            return np.busday_offset(start, self.n * count)
+
+        step = self.n
+
+        if self.name == "W":
+            step *= 7
```

Comment:
I'm still not clear what's going on here?

Reply:
From the docstring:

> Weekly frequency needs to be handled specially, since numpy counts the
> number of weeks since Thu Jan 1 1970 and uses Thursday and aligns the
> timestamp to Thursday. We therefore use daily frequency internally and
> align to Monday.